File: cane-cpp/cane-core/src/animation/sampling.cpp
#include "pose.hpp"
#include <limits>

namespace cane::animation {
namespace {
float& component(BoneLocal& local, std::size_t i) {
    switch (i) {
        case 0: return local.x; case 1: return local.y; case 2: return local.rotation_degrees;
        case 3: return local.scale_x; case 4: return local.scale_y; case 5: return local.shear_x_degrees;
        default: return local.shear_y_degrees;
    }
}
std::uint8_t color_byte(float value) { return static_cast<std::uint8_t>(std::floor(std::clamp(value, 0.0f, 1.0f) * 255.0f + .5f)); }
bool sampled_dark(const DiscreteChannel<bool>& channel, float time) {
    const auto& keys = channel.keys();
    const auto next = std::upper_bound(keys.begin(), keys.end(), time, [](float t, const DiscreteKey<bool>& key) { return t < key.time; });
    if (next == keys.begin()) return false;
    const auto& current = *std::prev(next);
    return current.value || (time != current.time && next != keys.end() && next->value);
}
std::uint32_t sequence_index(const DiscreteKey<SequenceSelection>& key, float time, std::size_t count) {
    const auto& value = key.value;
    if (value.mode == SequenceMode::hold || value.delay <= 0) return value.index;
    const double steps = std::floor((static_cast<double>(time) - key.time) / value.delay + .00001);
    const auto last = static_cast<std::uint32_t>(count - 1);
    const auto loop = [&] { return static_cast<std::uint32_t>(std::fmod(std::fmod(steps, static_cast<double>(count)) + value.index, static_cast<double>(count))); };
    const auto pingpong = [&](bool reverse) {
        if (count <= 1) return std::uint32_t{0};
        const double period = 2.0 * static_cast<double>(count) - 2;
        const auto phase = std::fmod(std::fmod(steps, period) + value.index + (reverse ? last : 0), period);
        return static_cast<std::uint32_t>(phase < static_cast<double>(count) ? phase : period - phase);
    };
    switch (value.mode) {
        case SequenceMode::once: return steps >= last - value.index ? last : value.index + static_cast<std::uint32_t>(steps);
        case SequenceMode::once_reverse: return steps >= last - value.index ? 0 : last - value.index - static_cast<std::uint32_t>(steps);
        case SequenceMode::loop: return loop();
        case SequenceMode::loop_reverse: return last - loop();
        case SequenceMode::pingpong: return pingpong(false);
        case SequenceMode::pingpong_reverse: return pingpong(true);
        case SequenceMode::hold: return value.index;
    }
    throw Error(ErrorCode::invalid_argument, "apply", "Unknown image sequence mode.");
}
float constrained_scalar(ConstraintKind kind, ConstraintProperty property, float value) {
    using P = ConstraintProperty;
    if (kind == ConstraintKind::physics) {
        if (property == P::mix || property == P::inertia || property == P::damping) return std::clamp(value, 0.0f, 1.0f);
        if (property == P::strength) return std::max(0.0f, value);
        if (property == P::mass && value <= 0) return std::numeric_limits<float>::denorm_min();
    } else if (kind == ConstraintKind::ik) {
        if (property == P::mix) return std::clamp(value, 0.0f, 1.0f);
        if (property == P::softness) return std::max(0.0f, value);
    } else if (kind == ConstraintKind::path && (property == P::mix_rotate || property == P::mix_x || property == P::mix_y)) return std::clamp(value, 0.0f, 1.0f);
    return value;
}
}
void Pose::apply(const Layer& layer) {
    if (!std::isfinite(layer.time) || !std::isfinite(layer.alpha)) throw Error(ErrorCode::non_finite, "apply", "Layer sample time and alpha must be finite.");
    const auto& clip = layer.clip; const auto& m = model();
    for (const auto& timeline : clip.bones) {
        auto& local = locals[timeline.index]; auto setup = data_.bones()[timeline.index].setup;
        for (std::size_t i = 0; i < timeline.channels.size(); ++i) {
            const bool angular = i == 2 || i >= 5;
            const auto sample = timeline.channels[i].sample(layer.time, angular, layer.stepped);
            if (sample) component(local, i) = blend_scalar(component(local, i), component(setup, i), *sample, layer.weight(PropertyDomain::bone, timeline.index, i), layer.additive, angular);
        }
        const auto inherit = timeline.inherit.sample(layer.time);
        if (inherit && layer.weight(PropertyDomain::bone, timeline.index, 7) >= .5f) modes[timeline.index] = *inherit;
    }
    for (const auto& timeline : clip.slots) {
        auto& slot = slots[timeline.index]; const auto& setup = m.slots[timeline.index];
        if (layer.attachments && layer.weight(PropertyDomain::slot, timeline.index, 7) >= .5f) {
            if (const auto key = timeline.attachment.sample(layer.time)) slot.key = *key;
        }
        const auto color_alpha = timeline.color[6].sample(layer.time, false, layer.stepped);
        if (color_alpha) {
            const auto blend_rgb = [&](detail::Rgb current, detail::Rgb base, std::size_t start) {
                for (std::size_t i = 0; i < current.size(); ++i) {
                    const auto value = timeline.color[start + i].sample(layer.time, false, layer.stepped);
                    if (value) current[i] = color_byte(blend_scalar(static_cast<float>(current[i]) / 255.0f, static_cast<float>(base[i]) / 255.0f, *value,
                        layer.weight(PropertyDomain::slot, timeline.index, start + i), layer.additive));
                }
                return current;
            };
            slot.color = blend_rgb(slot.color, setup.color, 0);
            bool dark_weight = false;
            for (std::size_t i = 3; i < 6; ++i) dark_weight = dark_weight || layer.weight(PropertyDomain::slot, timeline.index, i) != 0;
            if (dark_weight && (slot.dark || setup.dark || sampled_dark(timeline.dark_present, layer.time))) slot.dark = blend_rgb(slot.dark.value_or(detail::Rgb{}), setup.dark.value_or(detail::Rgb{}), 3);
        }
        const auto separate = timeline.alpha.sample(layer.time, false, layer.stepped);
        const auto alpha = !timeline.alpha.keys().empty() && color_alpha ? std::optional<float>(separate.value_or(setup.alpha)) : separate ? separate : color_alpha;
        if (alpha) slot.alpha = std::clamp(blend_scalar(slot.alpha, setup.alpha, *alpha, layer.weight(PropertyDomain::slot, timeline.index, 6), layer.additive), 0.0f, 1.0f);
    }
    for (const auto& timeline : clip.attachments) {
        auto& local = regions[timeline.index]; auto setup = m.attachments[timeline.index].local;
        for (std::size_t i = 0; i < timeline.region.size(); ++i) {
            const auto value = timeline.region[i].sample(layer.time, i == 2, layer.stepped);
            if (value) component(local, i) = blend_scalar(component(local, i), component(setup, i), *value, layer.weight(PropertyDomain::attachment, timeline.index, i), layer.additive, i == 2);
        }
        apply_deform(timeline, layer);
        if (layer.attachments && layer.weight(PropertyDomain::attachment, timeline.index, 6) >= .5f) {
            if (const auto key = timeline.sequence.sample_key(layer.time)) sequence_indices[timeline.index] = sequence_index(*key, layer.time, m.attachments[timeline.index].sequence_images.size());
        }
    }
    for (const auto& timeline : clip.constraints) {
        const auto apply_constraint = [&](std::size_t index) {
            const auto& definition = m.constraints[index]; auto& current = constraints[index];
            const auto domain = timeline.index ? PropertyDomain::constraint : PropertyDomain::physics_global;
            const auto target = timeline.index.value_or(0);
            for (const auto& channel : timeline.scalars) {
                const auto field = static_cast<std::size_t>(channel.property);
                if (!timeline.index && !std::get<detail::PhysicsDefinition>(definition.definition).globals[field]) continue;
                const auto value = channel.values.sample(layer.time, false, layer.stepped);
                if (value) {
                    const auto weight = layer.weight(domain, target, field);
                    if (weight != 0) current[channel.property] = constrained_scalar(definition.kind, channel.property,
                        blend_scalar(current[channel.property], definition.setup[channel.property], *value, weight, layer.additive && channel.property != ConstraintProperty::slider_time));
                }
            }
            for (const auto& channel : timeline.booleans) {
                const auto field = static_cast<std::size_t>(channel.property);
                const auto value = channel.values.sample(layer.time);
                if (value && layer.weight(domain, target, field) >= .5f) current.booleans.at(field - detail::scalar_parameter_count) = *value;
            }
        };
        if (timeline.index) apply_constraint(*timeline.index);
        else for (std::size_t i = 0; i < m.constraints.size(); ++i) if (m.constraints[i].kind == ConstraintKind::physics) apply_constraint(i);
    }
    if (layer.attachments && layer.weight(PropertyDomain::skin) >= .5f) {
        if (const auto skin = clip.skins.sample(layer.time)) { skins.clear(); if (*skin) skins.push_back(**skin); }
    }
    if (layer.draw_order && layer.weight(PropertyDomain::draw_order) >= .5f) {
        if (const auto sampled = clip.draw_order.sample(layer.time)) { order = *sampled; order_sampled = true; }
        for (const auto& folder : clip.draw_order_folders) {
            const auto key = folder.keys.sample(layer.time); if (!key) continue;
            std::fill(order_members_.begin(), order_members_.end(), false);
            for (const auto member : folder.members) order_members_[member] = true;
            std::size_t at = 0;
            for (auto& member : order) if (order_members_[member]) member = (*key)[at++];
            order_sampled = true;
        }
    }
}
}

File: cane-csharp/cane-core/Runtime/Animation/Sampling.cs
using System;
using System.Collections.Generic;
using Cane.Format;

namespace Cane.Animation
{
    internal sealed class Layer
    {
        internal Clip Clip = null!;
        internal float Time, Alpha;
        internal bool Additive, Attachments = true, DrawOrder = true, Stepped;
        internal Func<string, float>? PropertyAlpha;
        internal float Weight(string property) => PropertyAlpha == null ? Alpha : PropertyAlpha(property);
    }

    internal static class Sampling
    {
        private static readonly string[] RegionFields = { "x", "y", "rotation", "scaleX", "scaleY" };
        private static readonly string[] BooleanConstraintFields = { "bendPositive", "compress", "stretch" };
        internal static void Apply(Pose pose, Layer layer)
        {
            var data = pose.Data; var clip = layer.Clip;
            foreach (var t in clip.Bones)
            {
                BoneLocal current = pose.Locals[t.Index], setup = data.BoneData[t.Index].Setup;
                for (int component = 0; component < 7; component++)
                {
                    float? sample = t.Channels[component].Sample(layer.Time, component == 2 || component >= 5, layer.Stepped);
                    if (sample.HasValue) current = current.With(component, KeySampling.Blend(current.Get(component), setup.Get(component), sample.Value,
                        layer.Weight(t.Properties[component]), layer.Additive, component == 2 || component >= 5));
                }
                pose.Locals[t.Index] = current;
                Json inherit = KeySampling.Discrete(t.Inherit, layer.Time);
                if (!inherit.IsNull && layer.Weight(t.InheritProperty) >= 0.5f) pose.Modes[t.Index] = inherit.S("inherit");
            }
            foreach (Json t in clip.Raw["slotTimelines"].Items)
            {
                int index = data.SlotIndex[t.S("slotId")]; var setup = data.SlotData[index]; var slot = pose.Slots[index]; var channels = clip.Channels[t];
                Json attachment = KeySampling.Discrete(t["attachment"], layer.Time);
                if (!attachment.IsNull && layer.Attachments && layer.Weight("slot:" + index + ":attachment") >= 0.5f) slot.Key = attachment["attachmentId"].StringOrNull;
                float? colorAlpha = channels["color_alpha"].Sample(layer.Time, false, layer.Stepped);
                if (colorAlpha.HasValue)
                {
                    slot.Light = BlendRgb(channels, setup.Color, slot.Light, layer, "slot:" + index + ":", false);
                    bool dark = setup.Dark.HasValue || slot.Dark.HasValue;
                    if (!dark) dark = SampledDarkExists(t["color"], layer.Time);
                    if (dark) slot.Dark = BlendRgb(channels, setup.Dark ?? default, slot.Dark ?? default, layer, "slot:" + index + ":", true);
                }
                float? separate = channels["alpha"].Sample(layer.Time, false, layer.Stepped);
                float? alpha = channels["alpha"].HasKeys && colorAlpha.HasValue ? separate ?? setup.Alpha : separate ?? colorAlpha;
                if (alpha.HasValue) slot.Alpha = Numeric.Clamp(KeySampling.Blend(slot.Alpha, setup.Alpha, alpha.Value, layer.Weight("slot:" + index + ":alpha"), layer.Additive), 0, 1);
            }
            foreach (Json t in clip.Raw["attachmentTimelines"].Items)
            {
                int index = data.AttachmentIndex[t.S("attachmentId")]; var a = data.AttachmentData[index]; var channels = clip.Channels[t];
                BoneLocal setup = a.Setup, current = pose.Regions[index];
                string[] fields = RegionFields;
                for (int c = 0; c < 5; c++)
                {
                    float? value = channels[fields[c]].Sample(layer.Time, c == 2, layer.Stepped);
                    if (value.HasValue) current = current.With(c, KeySampling.Blend(current.Get(c), setup.Get(c), value.Value, layer.Weight("attachment:" + index + ":" + fields[c]), layer.Additive, c == 2));
                }
                pose.Regions[index] = current;
                if (t["deform"].ArrayOrEmpty.Count != 0 && channels.TryGetValue("d0", out Channel? deformChannel) && deformChannel.Sample(layer.Time).HasValue)
                {
                    int length = t["deform"][0]["vertices"].Items.Count; var sampled = new float[length];
                    for (int c = 0; c < length; c++) sampled[c] = channels["d" + c].Sample(layer.Time, false, layer.Stepped)!.Value;
                    bool weighted = a.GeometrySource.Raw["weights"].ArrayOrEmpty.Count != 0;
                    if (weighted && t.S("deformSpace", "vertexPositions") == "vertexPositions") sampled = CanonicalDeform(a.GeometrySource, sampled);
                    pose.Deforms.TryGetValue(a.DeformSourceId, out float[]? previous);
                    var values = new float[sampled.Length]; float alpha = layer.Weight("attachment:" + index + ":deform");
                    for (int c = 0; c < values.Length; c++)
                    { float baseValue = weighted ? 0 : a.GeometrySource.Vertices[c]; values[c] = KeySampling.Blend(previous == null ? baseValue : previous[c], baseValue, sampled[c], alpha, layer.Additive); }
                    pose.Deforms[a.DeformSourceId] = values;
                }
                Json sequence = KeySampling.Discrete(t["sequence"], layer.Time);
                if (!sequence.IsNull && layer.Attachments && layer.Weight("attachment:" + index + ":sequence") >= 0.5f)
                {
                    int count = a.Raw["sequence"]["imageIds"].Items.Count, value = sequence.I("index"); string mode = sequence.S("mode"); float delay = sequence.F("delay");
                    if (mode != "hold" && delay > 0)
                    {
                        double steps = Math.Floor(((double)layer.Time - sequence.F("time")) / delay + 0.00001);
                        // Reduce elapsed steps before converting to an index. Sequence playback has no signed-int time limit.
                        int LoopIndex() => (int)((steps % count + value) % count);
                        int Pingpong(bool reverse)
                        {
                            if (count <= 1) return 0;
                            double period = 2.0 * count - 2, offset = reverse ? count - 1 : 0;
                            double phase = (steps % period + value + offset) % period;
                            return (int)(phase < count ? phase : period - phase);
                        }
                        switch (mode)
                        {
                            case "once": value = steps >= count - 1 - value ? count - 1 : value + (int)steps; break;
                            case "loop": value = LoopIndex(); break;
                            case "pingpong": value = Pingpong(false); break;
                            case "onceReverse": value = steps >= count - 1 - value ? 0 : count - 1 - value - (int)steps; break;
                            case "loopReverse": value = count - 1 - LoopIndex(); break;
                            case "pingpongReverse": value = Pingpong(true); break;
                        }
                    }
                    pose.SequenceIndices[index] = Math.Min(Math.Max(value, 0), count - 1);
                }
            }
            foreach (Json t in clip.Raw["constraintTimelines"].Items)
            {
                bool wildcard = t.S("type") == "physics" && t.S("constraintId") == "*";
                for (int index = 0; index < data.Constraints.Length; index++)
                {
                Json current = pose.Constraints[index], setup = data.Constraints[index];
                if (wildcard ? setup.S("type") != "physics" : setup.S("id") != t.S("constraintId")) continue;
                foreach (var channel in clip.Channels[t])
                {
                    if (wildcard && !setup.B(channel.Key + "Global")) continue;
                    float? value = channel.Value.Sample(layer.Time, false, layer.Stepped); if (!value.HasValue) continue;
                    string field = channel.Key; float alpha = layer.Weight("constraint:" + t.S("constraintId") + ":" + field);
                    string targetField = field == "targetX" ? "x" : "y";
                    if (field == "targetX" || field == "targetY") Pose.Set(current["target"], targetField, KeySampling.Blend(current["target"].F(targetField), setup["target"].F(targetField), value.Value, alpha, layer.Additive));
                    else
                    {
                        string outputField = field == "sliderTime" ? "time" : field;
                        float sampled = KeySampling.Blend(ConstraintValue(current, outputField), ConstraintValue(setup, outputField), value.Value, alpha, field == "sliderTime" ? false : layer.Additive);
                        string kind = setup.S("type");
                        if (kind == "physics")
                        {
                            if (field == "mix" || field == "inertia" || field == "damping") sampled = Numeric.Clamp(sampled, 0, 1);
                            if (field == "strength") sampled = Math.Max(0, sampled);
                            if (field == "mass" && sampled <= 0) sampled = float.Epsilon;
                        }
                        else if (kind == "ik") { if (field == "mix") sampled = Numeric.Clamp(sampled, 0, 1); if (field == "softness") sampled = Math.Max(0, sampled); }
                        else if (kind == "path" && field.StartsWith("mix", StringComparison.Ordinal)) sampled = Numeric.Clamp(sampled, 0, 1);
                        Pose.Set(current, outputField, sampled);
                    }
                }
                foreach (var field in BooleanConstraintFields)
                {
                    Json key = KeySampling.Discrete(t["keys"], layer.Time, field);
                    if (!key.IsNull && layer.Weight("constraint:" + t.S("constraintId") + ":" + field) >= 0.5f) current.Members[field] = key[field];
                }
                }
            }
            if (layer.Attachments && layer.Weight("skin") >= 0.5f)
            {
                Json skin = KeySampling.Discrete(clip.Raw["skins"], layer.Time);
                if (!skin.IsNull) pose.Skins = skin["skinId"].IsNull ? Array.Empty<string>() : new[] { skin.S("skinId") };
            }
            if (layer.DrawOrder && layer.Weight("drawOrder") >= 0.5f)
            {
                Json order = KeySampling.Discrete(clip.Raw["drawOrder"], layer.Time);
                if (!order.IsNull)
                { pose.Order = new int[order["slotIds"].Items.Count]; for (int i = 0; i < pose.Order.Length; i++) pose.Order[i] = data.SlotIndex[order["slotIds"][i].String]; pose.OrderSampled = true; }
                foreach (Json folder in clip.Raw["drawOrderFolders"].Items)
                {
                    Json key = KeySampling.Discrete(folder["keys"], layer.Time); if (key.IsNull) continue;
                    var members = new HashSet<string>(folder["slotIds"].Strings()); int ordinal = 0;
                    for (int i = 0; i < pose.Order.Length; i++) if (members.Contains(data.SlotData[pose.Order[i]].Id)) pose.Order[i] = data.SlotIndex[key["slotIds"][ordinal++].String];
                    pose.OrderSampled = true;
                }
            }
        }
        private static bool SampledDarkExists(Json keys, float time)
        {
            IReadOnlyList<Json> rows = keys.ArrayOrEmpty; int low = 0, high = rows.Count;
            while (low < high) { int mid = low + (high - low) / 2; if (rows[mid].F("time") <= time) low = mid + 1; else high = mid; }
            if (low == 0) return false; Json current = rows[low - 1];
            return !current["darkColor"].IsNull || time != current.F("time") && low < rows.Count && !rows[low]["darkColor"].IsNull;
        }
        private static Rgb BlendRgb(Dictionary<string, Channel> channels, Rgb setup, Rgb current, Layer layer, string prefix, bool dark)
        {
            string channelPrefix = dark ? "dark_" : "color_";
            byte Blend(byte cur, byte bas, string field)
            {
                float value = channels[field].Sample(layer.Time, false, layer.Stepped) ?? cur / 255f;
                float blended = KeySampling.Blend(cur / 255f, bas / 255f, value, layer.Weight(prefix + field), layer.Additive);
                return (byte)Math.Floor(Numeric.Clamp(blended, 0, 1) * 255 + 0.5f);
            }
            return new Rgb(Blend(current.R, setup.R, channelPrefix + "r"), Blend(current.G, setup.G, channelPrefix + "g"), Blend(current.B, setup.B, channelPrefix + "b"));
        }
        internal static float ConstraintValue(Json c, string field)
        {
            if (!c[field].IsNull) return c.F(field);
            if (field.StartsWith("mix", StringComparison.Ordinal) || field == "inertia" || field == "damping" || field == "mass" || field == "timeScale") return 1;
            if (field == "strength") return 100; if (field == "fps") return 60; if (field == "limit") return 5000; return 0;
        }
        internal static float[] CanonicalDeform(AttachmentData source, float[] positions)
            => Geometry.Deformation.PositionsToOffsets(source, positions, "apply");
        internal static Affine ReadAffine(Json m) => new Affine(m.F("a"), m.F("b"), m.F("c"), m.F("d"), m.F("tx"), m.F("ty"));
    }
}
